src/utils/SpokeUtils.ts

```diff
@@ -0,0 +1,142 @@
+import { SpokePool } from "../typechain";
+
+/**
+ * Find the block range that contains the deposit ID. This is a binary search that searches for the block range
+ * that contains the deposit ID.
+ * @param targetDepositId The target deposit ID to search for.
+ * @param initLow The initial lower bound of the block range to search.
+ * @param initHigh The initial upper bound of the block range to search.
+ * @param maxSearches The maximum number of searches to perform. This is used to prevent infinite loops.
+ * @returns The block range that contains the deposit ID.
+ * @note  // We want to find the block range that satisfies these conditions:
+ *        // - the low block has deposit count <= targetDepositId
+ *        // - the high block has a deposit count > targetDepositId.
+ *        // This way the caller can search for a FundsDeposited event between [low, high] that will always
+ *        // contain the event emitted when deposit ID was incremented to targetDepositId + 1. This is the same transaction
+ *        // where the deposit with deposit ID = targetDepositId was created.
+ */
+export async function getBlockRangeForDepositId(
+  targetDepositId: number,
+  initLow: number,
+  initHigh: number,
+  maxSearches: number,
+  spokePool: SpokePool,
+  deploymentBlock = 0
+): Promise<{
+  low: number;
+  high: number;
+}> {
+  // Get the most recent block from the spoke pool contract.
+  const mostRecentBlockNumber = await spokePool.provider.getBlockNumber();
+  // Set the initial high block to the most recent block number or the initial high block, whichever is smaller.
+  initHigh = Math.min(initHigh, mostRecentBlockNumber);
+
+  // Define a mapping of block numbers to deposit IDs. This is used to cache the deposit ID at a block number
+  // so we don't need to make an eth_call request to get the deposit ID at a block number more than once.
+  const queriedIds: Record<number, number> = {};
+
+  // Define a llambda function to get the deposit ID at a block number. This function will first check the
+  // queriedIds cache to see if the deposit ID at the block number has already been queried. If not, it will
+  // make an eth_call request to get the deposit ID at the block number. It will then cache the deposit ID
+  // in the queriedIds cache.
```

Comment:
Just my 2c - I think the function is pretty self-explanatory, so 4 lines of comments to a 6 line function might be a bit verbose.

Reply:
I want to push on this and keep the documentation specifically because it's an anonymous function that uses the scope of the variable above it.